### PR TITLE
Add DB trigger to populate SIGMA DAO table

### PR DESCRIPTION
**Pull request by rmeena840**
Add DB trigger to populate SIGMA DAO table

https://www.pivotaltracker.com/story/show/182046917

### CODE REVIEW

File: infrastructure/setup_sigma_dao.sql [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    END LOOP;
+	RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- create a trigger (update_app)
+-- Event: INSERT on publoc.txn relation
+CREATE TRIGGER update_app
+AFTER INSERT ON public.txn FOR EACH ROW
+EXECUTE FUNCTION update_app_info();
```
Comment (robert-zaremba) [inline infrastructure/setup_sigma_dao.sql:42]:
```suggestion
EXECUTE FUNCTION sigmadao_trigger_fn();
```

File: infrastructure/setup_sigma_dao.sql [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+			EXIT;
+		END IF;
+    END LOOP;
+	RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- create a trigger (update_app)
+-- Event: INSERT on publoc.txn relation
+CREATE TRIGGER update_app
```
Comment (robert-zaremba) [inline infrastructure/setup_sigma_dao.sql:40]:
```suggestion
CREATE TRIGGER sigmadao_trigger
```

File: infrastructure/setup_sigma_dao.sql [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	FOR i IN SELECT txn FROM txn WHERE asset = NEW.asset AND typeenum=6 LOOP
+		IF i.txn::jsonb -> 'dt'->'lg' ?& ARRAY[log_sigma] THEN
+			-- store token id from db to above decalred variable token_id
+			SELECT index INTO token_id FROM public.asset ORDER BY index DESC LIMIT 1;
+			-- store token name from db to above declared variable token_name
+			SELECT (SELECT params FROM public.asset WHERE index = token_id)::jsonb -> 'an' INTO token_name;
+			-- insert values in sigma_daos table
+			INSERT INTO public.sigma_daos (
+			app_id, app_params, token_id, token_name)
+			VALUES (NEW.asset, NEW.txn, token_id, token_name);
```
Comment (robert-zaremba) [inline infrastructure/setup_sigma_dao.sql:30]: I think this is not correct
1. you are assigning `NEW.asset` as `app_id`.
2. let's remove `token_name`.

File: infrastructure/setup_sigma_dao.sql [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+DECLARE
+    token_id bigint;
+	token_name varchar(20);
+	log_sigma text := 'U2lnbWFEQU8gY3JlYXRlZA==';
+	i record;
+BEGIN
+	FOR i IN SELECT txn FROM txn WHERE asset = NEW.asset AND typeenum=6 LOOP
+		IF i.txn::jsonb -> 'dt'->'lg' ?& ARRAY[log_sigma] THEN
+			-- store token id from db to above decalred variable token_id
+			SELECT index INTO token_id FROM public.asset ORDER BY index DESC LIMIT 1;
```
Comment (robert-zaremba) [inline infrastructure/setup_sigma_dao.sql:24]: the DAO token ID is not the most recent asset. It  can be any asset. You should read it from the app global state or a log.

**Review by robert-zaremba — CHANGES_REQUESTED** [review]
1. Assigning DAO token ID must be fixed
2. I think `NEW.asset` is wrongly used - we need app id.

File: infrastructure/setup_sigma_dao.sql (hunk 3 above)
Comment (rmeena840) [reply infrastructure/setup_sigma_dao.sql]: 1. We disccused about `app_id` in general
2. Ok. Done

File: infrastructure/setup_sigma_dao.sql [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+RETURNS TRIGGER
+AS $$
+DECLARE
+    token_id bigint;
+	token_name varchar(20);
+	log_sigma text := 'U2lnbWFEQU8gY3JlYXRlZA==';
+	i record;
+BEGIN
+	FOR i IN SELECT txn FROM txn WHERE asset = NEW.asset AND typeenum=6 LOOP
+		IF i.txn::jsonb -> 'dt'->'lg' ?& ARRAY[log_sigma] THEN
```
[the remark this answers is not in the document]
Comment (rmeena840) [reply infrastructure/setup_sigma_dao.sql:24]: Done

File: infrastructure/Makefile [hunk 6]
```diff
@@ -153,6 +155,10 @@ recreate-indexer-db:
 	@sudo -u postgres createdb $(POSTGRES_DBNAME)
 	@sudo -u postgres psql -c "grant all privileges on database $(POSTGRES_DBNAME) to $(POSTGRES_UNAME)"
 
+# Setups sigma_daos table. Run this after running start-indexer script
+setup-sigma-daos:
+	@psql -U $(POSTGRES_UNAME) -d $(POSTGRES_DBNAME) -a -f $(SIGMA_DAO_SQL_FILE_PATH)
```
[the remark this answers is not in the document]
Comment (rmeena840) [reply infrastructure/Makefile]: Done

File: infrastructure/setup_sigma_dao.sql (hunk 4 above)
Comment (rmeena840) [reply infrastructure/setup_sigma_dao.sql]: Yes, I know this. Currently, we are not storing asset id in a global state or log so it's hard to determine which asset was along with the app. This issue will be resolved once the asset has been moved to a global state with the hash code task. https://www.pivotaltracker.com/story/show/182178388

Comment (robert-zaremba) [reply infrastructure/setup_sigma_dao.sql]: please add a subtask there and TODO in the SQL code here

**Review by robert-zaremba — APPROVED** [review]
Pending adding TODO from my other comment.

File: infrastructure/setup_sigma_dao.sql (hunk 4 above)
Comment (rmeena840) [reply infrastructure/setup_sigma_dao.sql]: Done

**Pull request merged**